### PR TITLE
fix: robust websocket disconnect, nonblocking usage stats

### LINKED ISSUE #1622: WebSocketDisconnect (kernel not found) when using marimo edit
### Describe the bug

Hi all, I'm using `conda 4.8.1` and create a `Python 3.9.18` environment, on MacOS 10.15.7.
After run `pip install marimo`, run `marimo tutorial intro` it works and show the notebook.

But if I run `marimo edit`, it open explorer then show `kernel not found`, the terminal outputs
```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 332, in asgi_send
    await self.send(data)  # type: ignore[arg-type]
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/websockets/legacy/protocol.py", line 647, in send
    await self.write_frame(True, opcode, data)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/websockets/legacy/protocol.py", line 1209, in write_frame
    await self.drain()
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/websockets/legacy/protocol.py", line 1198, in drain
    await self.ensure_open()
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/websockets/legacy/protocol.py", line 939, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: no close frame received or sent

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/websockets.py", line 97, in send
    await self._send(message)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/_exception_handler.py", line 50, in sender
    await send(message)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/middleware/sessions.py", line 83, in send_wrapper
    await send(message)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 345, in asgi_send
    raise ClientDisconnected from exc
uvicorn.protocols.utils.ClientDisconnected

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 244, in run_asgi
    result = await self.app(self.scope, self.asgi_receive, self.asgi_send)  # type: ignore[func-returns-value]
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
    return await self.app(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/middleware/errors.py", line 151, in __call__
    await self.app(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/marimo/_server/api/auth.py", line 215, in __call__
    return await super().__call__(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/middleware/sessions.py", line 85, in __call__
    await self.app(scope, receive, send_wrapper)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/middleware/authentication.py", line 49, in __call__
    await self.app(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/middleware/cors.py", line 77, in __call__
    await self.app(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/marimo/_server/api/middleware.py", line 64, in __call__
    return await self.app(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/routing.py", line 485, in handle
    await self.app(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/routing.py", line 373, in handle
    await self.app(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/routing.py", line 96, in app
    await wrap_app_handling_exceptions(app, session)(scope, receive, send)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/routing.py", line 94, in app
    await func(session)
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/marimo/_server/api/endpoints/ws.py", line 76, in websocket_endpoint
    await WebsocketHandler(
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/marimo/_server/api/endpoints/ws.py", line 386, in start
    await self.future
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/marimo/_server/api/endpoints/ws.py", line 318, in listen_for_messages
    await self.websocket.send_text(
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/websockets.py", line 188, in send_text
    await self.send({"type": "websocket.send", "text": data})
  File "/Users/art9/anaconda3/envs/py39/lib/python3.9/site-packages/starlette/websockets.py", line 100, in send
    raise WebSocketDisconnect(code=1006)
starlette.websockets.WebSocketDisconnect
        Are you sure you want to quit? (y/n): ^C
```

I tried copy the intro.py to workspace folder and run `marimo edit intro.py`, it does not work.

I am confused about why `tutorial intro` command works, but `edit ` command failed. Is there a difference and maybe related to my python env?

### Environment

{
  "marimo": "0.6.19",
  "OS": "Darwin",
  "OS Version": "19.6.0",
  "Processor": "i386",
  "Python Version": "3.9.18",
  "Binaries": {
    "Browser": "115.0.5790.170",
    "Node": "v18.12.1"
  },
  "Requirements": {
    "click": "8.1.7",
    "importlib-resources": "6.4.0",
    "jedi": "0.19.1",
    "markdown": "3.6",
    "pymdown-extensions": "10.8.1",
    "pygments": "2.17.2",
    "tomlkit": "0.12.5",
    "uvicorn": "0.30.1",
    "starlette": "0.37.2",
    "websockets": "12.0",
    "typing-extensions": "4.10.0",
    "black": "24.2.0"
  }
}

### Code to reproduce

_No response_

### LINKED ISSUE #1653: marimo server becomes unusable when several notebooks are open
### Describe the bug

On macOS, after opening many notebooks, the marimo server can become unusable — notebooks refuse to open, incorrectly claiming that another browser tab is already connected. I don't run into this issue on Ubuntu.

This could be related to #1622.

The following errors are repeatedly emitted in the console:

```
ERROR:    Exception in ASGI application                                                                                                              
Traceback (most recent call last):                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 332, in asgi_send            
    await self.send(data)  # type: ignore[arg-type]                                                                                                  
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/websockets/legacy/protocol.py", line 635, in send                                   
    await self.ensure_open()                                                                                                                         
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/websockets/legacy/protocol.py", line 939, in ensure_open                            
    raise self.connection_closed_exc()                                                                                                               
websockets.exceptions.ConnectionClosedError: no close frame received or sent                                                                         
                                                                                                                                                     
The above exception was the direct cause of the following exception:                                                                                 
                                                                                                                                                     
Traceback (most recent call last):                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/websockets.py", line 97, in send                                          
    await self._send(message)                                                                                                                        
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 50, in sender                                
    await send(message)                                                                                                                              
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/sessions.py", line 83, in send_wrapper                         
    await send(message)                                                                                                                              
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 345, in asgi_send            
    raise ClientDisconnected from exc                                                                                                                
uvicorn.protocols.utils.ClientDisconnected                                                                                                           
                                                                                                                                                     
During handling of the above exception, another exception occurred:                                                                                  
                                                                                                                                                     
Traceback (most recent call last):                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 244, in run_asgi                                                  
    result = await self.app(self.scope, self.asgi_receive, self.asgi_send)  # type: ignore[func-returns-value]                                                                             
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__                                                               
    return await self.app(scope, receive, send)                                                                                                                                            
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/applications.py", line 123, in __call__                                                                        
    await self.middleware_stack(scope, receive, send)                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/errors.py", line 151, in __call__                                                                   
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/auth.py", line 215, in __call__                                                                       
    return await super().__call__(scope, receive, send)                                                                                                                                    
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/sessions.py", line 85, in __call__                                                                  
    await self.app(scope, receive, send_wrapper)                                                                                                                                           
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/authentication.py", line 49, in __call__                                                            
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/cors.py", line 77, in __call__                                                                      
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/middleware.py", line 64, in __call__                                                                  
    return await self.app(scope, receive, send)                                                                                                                                            
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 65, in __call__                                                                
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)                                                                                                               
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app                                                                
    raise exc                                                                                                                                                                              
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app                                                                
    await app(scope, receive, sender)                                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 756, in __call__                                                                             
    await self.middleware_stack(scope, receive, send)                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 776, in app                                                                                  
    await route.handle(scope, receive, send)                                                                                                                                               
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 485, in handle                                                                               
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 756, in __call__                                                                             
    await self.middleware_stack(scope, receive, send)                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 776, in app                                                                                  
    await route.handle(scope, receive, send)                                                                                                                                               
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 373, in handle                                                                               
    await self.app(scope, receive, send)                                                                                                                                                                      [413/896]
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 96, in app                                                                                                               
    await wrap_app_handling_exceptions(app, session)(scope, receive, send)                                                                                                                                             
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app                                                                                            
    raise exc                                                                                                                                                                                  
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app                                                                                            
    await app(scope, receive, sender)                                                                                                                                                                                  
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 94, in app                                                                                                               
    await func(session)                                                                                                                                                                                                
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/endpoints/ws.py", line 76, in websocket_endpoint                                                                                  
    await WebsocketHandler(                                                                                                                                                                                            
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/endpoints/ws.py", line 386, in start                                                                                              
    await self.future                                                                                                                  
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/endpoints/ws.py", line 318, in listen_for_messages                                                                                
    await self.websocket.send_text(                                                                                                                                                                                    
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/websockets.py", line 188, in send_text                                                                                                     
    await self.send({"type": "websocket.send", "text": data})                                                                                                                                                          
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/websockets.py", line 100, in send                                                                                                          
    raise WebSocketDisconnect(code=1006)                                                                                                                                                                               
starlette.websockets.WebSocketDisconnect                                                                      
```

```
ERROR:    Exception in ASGI application                                                                                                                                                    
Traceback (most recent call last):                                                                                                                                                         
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 244, in run_asgi                                                  
    result = await self.app(self.scope, self.asgi_receive, self.asgi_send)  # type: ignore[func-returns-value]                                                                             
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__                                                               
    return await self.app(scope, receive, send)                                                                                                                                            
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/applications.py", line 123, in __call__                                                                        
    await self.middleware_stack(scope, receive, send)                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/errors.py", line 151, in __call__                                                                   
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/auth.py", line 215, in __call__                                                                       
    return await super().__call__(scope, receive, send)                                                                                                                                    
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/sessions.py", line 85, in __call__                                                                  
    await self.app(scope, receive, send_wrapper)                                                                                                                                           
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/authentication.py", line 49, in __call__                                                            
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/cors.py", line 77, in __call__                                                                      
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/middleware.py", line 64, in __call__                                                                  
    return await self.app(scope, receive, send)                                                                                                                                            
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 65, in __call__                                                                
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)                                                                                                               
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app                                                                
    raise exc                                                                                                                                                                              
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app                                                                
    await app(scope, receive, sender)                                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 756, in __call__                                                                             
    await self.middleware_stack(scope, receive, send)                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 776, in app                                                                                  
    await route.handle(scope, receive, send)                                                                                                                                               
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 485, in handle                                                                               
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 756, in __call__                                                                             
    await self.middleware_stack(scope, receive, send)                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 776, in app                                                                                  
    await route.handle(scope, receive, send)                                                                                                                                               
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 373, in handle                                                                               
    await self.app(scope, receive, send)                                                                                                                                                   
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 96, in app                                                                                   
    await wrap_app_handling_exceptions(app, session)(scope, receive, send)                                                                                                                 
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app                                                                
    raise exc                                                                                                                                                                              
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app                                                                
    await app(scope, receive, sender)                                                                                                                                                      
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/routing.py", line 94, in app                                                                                   
    await func(session)                                                                                                                                                                    
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/endpoints/ws.py", line 76, in websocket_endpoint                                                      
    await WebsocketHandler(                                                                                                                                                                
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/marimo/_server/api/endpoints/ws.py", line 249, in start                                                                  
    await self.websocket.close(                                                                                                                                                            
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/websockets.py", line 203, in close                                                                             
    await self.send(                                                                                                                                                                         File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/websockets.py", line 97, in send                                          
    await self._send(message)                                                                                                                                                                File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/_exception_handler.py", line 50, in sender                                
    await send(message)                                                                                                                                                                    
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/starlette/middleware/sessions.py", line 83, in send_wrapper                                                              
    await send(message)                                                                                                                                                                    
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 338, in asgi_send            
    await self.close(code, reason)                                                                                                                                                         
  File "/Users/akshay/envs/playground/lib/python3.9/site-packages/websockets/legacy/protocol.py", line 788, in close                                  
    await self.transfer_data_task                                                                                                                                                          
AttributeError: 'WebSocketProtocol' object has no attribute 'transfer_data_task'                                                                                                           
ERROR:    Exception in ASGI application                             
```

### Environment

```
{
  "marimo": "0.6.19",
  "OS": "Darwin",
  "OS Version": "22.5.0",
  "Processor": "arm",
  "Python Version": "3.9.19",
  "Binaries": {
    "Browser": "126.0.6478.62",
    "Node": "v21.5.0"
  },
  "Requirements": {
    "click": "8.1.7",
    "importlib-resources": "5.10.2",
    "jedi": "0.19.1",
    "markdown": "3.6",
    "pymdown-extensions": "10.8.1",
    "pygments": "2.18.0",
    "tomlkit": "0.12.5",
    "uvicorn": "0.30.1",
    "starlette": "0.37.2",
    "websockets": "12.0",
    "typing-extensions": "4.12.2",
    "black": "24.4.2"
  }
}
```

### Code to reproduce

1. Run marimo edit in a directory with several notebooks
2. Open around 10 notebooks.

### PR DESCRIPTION
**Pull request by akshayka**
1. When multiple notebooks are open in marimo edit, we sometimes get unexpected `WebSocketDisconnect`s. This change makes the server more robust to those errors — one can at least continue working with notebooks that failed to connect by refreshing —  without solving the root problem.

2. Make usage endpoint's CPU measurement nonblocking. This was slowing down the server.

Fixes #1653 
Fixes #1622

### CODE REVIEW

File: marimo/_server/sessions.py [hunk 1]
```diff
@@ -366,13 +366,18 @@ def put_input(self, text: str) -> None:
...
+                "Expecting a session consumer to pause, but none found"
+            )
+            return
+
         LOGGER.debug("Disconnecting session consumer")
-        self.session_consumer.on_stop()
-        self.unsubscribe_consumer()
-        self.session_consumer = None
+        try:
+            self.session_consumer.on_stop()
```
Comment (akshayka) [inline marimo/_server/sessions.py:377]: This sometimes fails with errors deep in the guts of `websockets`. Finally block makes sure the consumer is removed anyway, so users can still work with the notebook by refreshing

File: marimo/_server/api/endpoints/ws.py [hunk 2]
```diff
@@ -323,64 +366,50 @@ async def listen_for_messages() -> None:
...
                     LOGGER.error("Message: %s", data)
+                    continue
+
+                try:
+                    await self.websocket.send_text(text)
+                except WebSocketDisconnect as e:
+                    self._on_disconnect(
+                        e,
+                        cleanup_fn=lambda: listen_for_disconnect_task.cancel(),
+                    )
```
Comment (akshayka) [inline marimo/_server/api/endpoints/ws.py:390]: Abnormal disconnects can happen when sending too — and it seems like these can race with the `listen_for_disconnect` task?

File: marimo/_server/api/endpoints/ws.py [hunk 3]
```diff
@@ -323,64 +366,50 @@ async def listen_for_messages() -> None:
...
+                except RuntimeError as e:
+                    # Starlette can raise a runtime error if a message is sent
+                    # when the socket is closed. In case the disconnection
+                    # error hasn't made its way to listen_for_disconnect, do
```
Comment (akshayka) [inline marimo/_server/api/endpoints/ws.py:394]: I am not really sure if this case is needed. I occasionally saw `RuntimeError`s raised on `send_text` saying the socket was already closed

**Comment by mscolnick** [comment]
can we wait till https://github.com/marimo-team/marimo/pull/1639 merges? it will simplify this and otherwise cause a few merge conflicts

**Review by mscolnick — APPROVED** [review]

File: frontend/src/core/websocket/useWebSocket.tsx [hunk 4]
```diff
@@ -35,6 +35,9 @@ export function useWebSocket(options: UseWebSocketOptions) {
             maxRetries: 10,
             debug: false,
             startClosed: true,
+            // long timeout -- the server can become slow when many notebooks
+            // are open.
+            connectionTimeout: 60_000,
```
Comment (mscolnick) [inline frontend/src/core/websocket/useWebSocket.tsx:40]: looks like the default is 4s (just to have it handy here) https://github.com/partykit/partykit/blob/50715a06242732ac828993d2c15d63d30846efb2/packages/partysocket/src/ws.ts#L120

**Review by mscolnick — APPROVED** [review]

**Pull request merged**